README
		 Periodic - a repeating event library
		 ------------------------------------
		  David Shaw <dshaw@jabberwocky.com>

There isn't a lot that needs to be said about this library.
Basically, it acts like a "cron" inside a program.  You give it a list
of functions that need to be run and how often you want them to run,
and periodic schedules and runs them as needed.  I've written the
rough equivalent of this code a number of times in various jobs, and
it became time to just write a nice robust version of it, give it a
clean API, and release it as a library.  If nothing else, now I don't
have to write it over again.

See the periodic.h header for instructions, and the examples directory
for sample code.


NEWS
Noteworthy changes in version 0.6 (unreleased)
----------------------------------------------

    * Initial release.


examples/periodic-test.c
#include <config.h>
#include <stdio.h>
#include <unistd.h>
#include <time.h>
#include <periodic.h>

static void
one(void *foo)
{
  time_t now=time(NULL);

  printf("ONE SECOND EVENT: It is %.24s and my arg is %p\n",ctime(&now),foo);
}

static void
three(void *foo)
{
  time_t now=time(NULL);

  printf("THREE SECOND EVENT: It is %.24s and my arg is %p\n",ctime(&now),foo);
}

static void
five(void *foo)
{
  time_t now=time(NULL);

  printf("FIVE SECOND EVENT: It is %.24s and my arg is %p\n",ctime(&now),foo);
}

static void
oneshot(void *foo)
{
  time_t now=time(NULL);

  printf("ONESHOT EVENT: It is %.24s and my arg is %p\n",ctime(&now),foo);
}

static void
timewarp(void *foo)
{
  time_t now=time(NULL);

  printf("TIMEWARP: It is %.24s and my arg is %p\n",ctime(&now),foo);
}

int
main(int argc,char *argv[])
{
  struct periodic_event_t *event1;

  (void)argc;
  (void)argv;

  periodic_timewarp(1,0,timewarp,NULL);

  event1=periodic_add(1,0,one,(void *)0x1234);
  periodic_add(3,0,three,(void *)0x5678);
  periodic_add(5,0,five,(void *)0x5678);

  periodic_add(5,PERIODIC_DELAY|PERIODIC_ONESHOT,oneshot,(void *)0x5678);

  periodic_start(PERIODIC_DEBUG);

  sleep(5);

  periodic_remove(event1);

  pause();

  /* Never reached */
  return 0;
}


include/periodic.h
/*
  periodic - a library for repeating periodic events
  Copyright (C) 2008, 2009 David Shaw, <dshaw@jabberwocky.com>

  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
  02110-1301, USA
*/

#ifndef _PERIODIC_H_
#define _PERIODIC_H_

#include <time.h>

#ifdef __cplusplus
extern "C" {
#endif

/*
  Call this function to add an event to the queue.  Interval is how
  often (in seconds) you want this function to run.  The func is a
  function with the following prototype:

    void my_function(void *arg)

  "now" is the current time.  "arg" is passed as the last argument to
  periodic_event, and is passed through to the func.

  The flags field can be any number of the PERIODIC_xxx flags above,
  ORed together.  DELAY means to start the event after the interval
  (that is, a 5-second event happens after 5 seconds, rather than
  immediately after adding it and then again after 5 seconds).
  ONESHOT means to trigger this event once and never again.

  This function returns a pointer to a struct periodic_event_t on
  success, or NULL on failure (and sets errno).
 */

#define PERIODIC_DELAY   1
#define PERIODIC_ONESHOT 2

struct periodic_event_t *periodic_add(unsigned int interval,unsigned int flags,
				      void (*func)(void *),void *arg);

/*
  Call this function a remove an event from the queue.  Pass the
  struct periodic_event_t pointer returned by periodic_add().  Returns
  0 for success and -1 for failure (and sets errno).
 */

int periodic_remove(struct periodic_event_t *remove);

/*
  Call this function to start the periodic events.  Concurrency is how
  many periodic events can happen at once.  If multiple events happen
  at the same moment and concurrency is not high enough, the events
  will happen one after another rather than concurrently.  The
  NORETURN flag makes periodic_start() not return - the calling thread
  becomes one of the pool used to run periodic events.
 */

#define PERIODIC_DEBUG    1
#define PERIODIC_NORETURN 2

int periodic_start(unsigned int flags);

/*
  Stops calling periodic events.  Note that this does not remove any
  events from the queue.  PERIODIC_WAIT means to wait for all threads
  to exit.  Otherwise, the threads are signalled to exit and may exit
  after periodic_stop() returns.
 */

#define PERIODIC_WAIT 1

int periodic_stop(unsigned int flags);

/*
  A common problem with programs that use any sort of repeating event
  is what to do in the case of a time warp (i.e. a clock that jumps in
  either direction).  For example, if it is 15:00:00 and you have an
  event scheduled for 15:00:08, and the clock jumps back 5 hours to
  10:00:00, then the event that should have happened in 8 seconds
  won't happen for 5 hours and 8 seconds.  This function tells the
  periodic library to detect and handle such and event (by fixing the
  schedule so that events happen on time).  "interval" is how often to
  check for a timewarp, and "warptime" is how much of a jump is
  allowed before the event schedule is fixed.

  You may also pass a pointer to timewarp handler function that will
  be called so your process can do its own timewarp handling in
  addition to the above.  You may pass NULL for this if you do not
  need your own handler function.  The prototype for this handler is
  the same as a periodic event.

  Note that periodic will use a monotonic clock on those platforms
  that support it.  This makes it immune from timewarps for its own
  events (i.e. things added via periodic_add), but you may still need
  this function for other purposes if your program uses wall-clock
  time.
*/

int periodic_timewarp(unsigned int interval,unsigned int warptime,
		      void (*func)(void *),void *arg);

#ifdef __cplusplus
}
#endif

#endif /* !_PERIODIC_H_ */


lib/periodic.c
/*
  periodic - a library for repeating periodic events
  Copyright (C) 2008, 2009, 2011, 2015 David Shaw, <dshaw@jabberwocky.com>

  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
  02110-1301, USA
*/

#include <config.h>
#include <pthread.h>
#include <time.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <stdio.h>
#include <periodic.h>

static struct periodic_event_t
{
  unsigned int interval;
  time_t next_occurance;
  time_t last_start;
  unsigned int elapsed;
  unsigned int count;
  void (*func)(void *);
  void *arg;
  struct
  {
    unsigned int oneshot:1;
  } flags;
  struct periodic_event_t *next;  
} *events=NULL;

static pthread_mutex_t event_lock=PTHREAD_MUTEX_INITIALIZER;
static pthread_cond_t event_cond;
static pthread_mutex_t thread_lock=PTHREAD_MUTEX_INITIALIZER;
static unsigned int num_threads;
static unsigned int idle_threads;
static pthread_t *threads;
static pthread_t timewarp;
static unsigned int timewarp_interval;
static unsigned int timewarp_warptime;
static void (*timewarp_func)(void *);
static void *timewarp_arg;
static unsigned int global_flags;

#define debug(_fmt,_vargs...) do {if(global_flags&PERIODIC_DEBUG) fprintf(stderr,"Periodic: "_fmt,##_vargs); } while(0)

static void *periodic_thread(void *foo);

/* Do the best we can here.  We want to return time from the monotonic
   clock, but it might not exist on this platform. */
static time_t
gettime(void)
{
  /* Only use the monotonic clock if we can use it in both gettime and
     the pthreads condition variable. */
#if defined(HAVE_CLOCK_GETTIME) && defined(HAVE_PTHREAD_CONDATTR_SETCLOCK)

#ifndef CLOCK_MONOTONIC
#define CLOCK_MONOTONIC CLOCK_REALTIME
#endif

  struct timespec ts;

  if(clock_gettime(CLOCK_MONOTONIC,&ts)==0)
    return ts.tv_sec;
  else
#endif /* HAVE_CLOCK_GETTIME && HAVE_PTHREAD_CONDATTR_SETCLOCK */
    return time(NULL);
}

/* Called with event_lock locked */
static void
attach(void *e)
{
  struct periodic_event_t *event=e;

  event->next=events;
  events=event;
}

static void
enqueue(void *e)
{
  struct periodic_event_t *event=e;

  pthread_mutex_lock(&event_lock);

  if(event->flags.oneshot)
    free(event);
  else
    {
      time_t now=gettime();

      /* Reschedule it */
      event->next_occurance=now+event->interval;
      event->elapsed+=now-event->last_start;
      event->count++;

      event->next=events;
      events=event;
    }

  pthread_mutex_unlock(&event_lock);
}

static void
unlocker(void *foo)
{
  pthread_mutex_unlock(&event_lock);
}

/* Called with thread_lock locked */
static int
make_new_thread(void)
{
  pthread_t *new_threads;
  int err;

  debug("Making a new thread (will have %d)\n",num_threads+1);

  new_threads=realloc(threads,sizeof(pthread_t)*(num_threads+1));
  if(!new_threads)
    return ENOMEM;

  threads=new_threads;

  err=pthread_create(&threads[num_threads],NULL,periodic_thread,NULL);
  if(err==0)
    {
      num_threads++;
      return 0;
    }
  else
    {
      /* We don't even try to lower the size of the threads array.  At
	 worst, we're wasting sizeof(pthread_t). */
      return err;
    }
}

/* count is the number of items that should be happening right now.
   If it is more than 1, and we don't have a spare thread, then
   someone is going to wait.  This might be a problem if the event
   takes a while.  after_occurance is the next time an event will fire
   off.  It might be in the past if we have events that take a while.
   It also might be equal to the current time if count>1. */

static struct periodic_event_t *
dequeue(void)
{  
  struct periodic_event_t *last_event=NULL,*next_event;
  time_t now,after_occurance;
  int err;

  debug("%s","\n");

  pthread_mutex_lock(&thread_lock);

  idle_threads++;

  pthread_mutex_unlock(&thread_lock);

  pthread_mutex_lock(&event_lock);

  for(;;)
    {
      struct periodic_event_t *event,*last=NULL;
      time_t next_occurance=0x7FFFFFFF; /* 2038 */

      after_occurance=0x7FFFFFFF; /* 2038 */

      next_event=NULL;

      /* Find the next event to occur */
      for(event=events;event;last=event,event=event->next)
	{
	  if(event->next_occurance<next_occurance)
	    {
	      next_occurance=event->next_occurance;
	      if(next_event)
		after_occurance=next_event->next_occurance;

	      next_event=event;
	      last_event=last;
	    }
	  else if(event->next_occurance<after_occurance)
	    after_occurance=event->next_occurance;
	}

      /* Now wait for the event time to arrive. */

      if(next_event)
	{
	  struct timespec timeout;

	  /* Remove the event from the list */
	  if(last_event)
	    last_event->next=next_event->next;
	  else
	    events=next_event->next;

	  next_event->next=NULL;

	  timeout.tv_sec=next_occurance;
	  timeout.tv_nsec=0;

	  /* Is our own event going to happen again before the next
	     event? */
	  if(next_event->next_occurance+next_event->interval<after_occurance)
	    after_occurance=next_event->next_occurance+next_event->interval;

	  pthread_cleanup_push(attach,next_event);
	  pthread_cleanup_push(unlocker,NULL);

	  err=pthread_cond_timedwait(&event_cond,&event_lock,&timeout);

	  pthread_cleanup_pop(0);
	  pthread_cleanup_pop(0);

	  /* A new event showed up, so recalculate (and put back the
	     event we were waiting on).  Also double check to cover
	     for minor clock jitter. */

	  now=gettime();

	  if(err==0 || now<next_occurance)
	    {
	      next_event->next=events;
	      events=next_event;
	      continue;
	    }
	}
      else
	{
	  /* Wait forever */

	  pthread_cleanup_push(unlocker,NULL);

	  pthread_cond_wait(&event_cond,&event_lock);

	  pthread_cleanup_pop(0);

	  /* A new event showed up, so recalculate. */
	  continue;
	}

      /* If we get to here, we have an event, and its time has been
	 reached. */

      break;
    }

  /* Now we have an event, and we know it needs to start now.  We also
     know how long on average this event takes to run.  Does the next
     event start before this one is expected to finish?  If so, and we
     don't have a spare thread handy, then we may have a problem. */

  pthread_mutex_lock(&thread_lock);

  idle_threads--;

  pthread_cleanup_push(attach,next_event);

  if(idle_threads==0
     && next_event->count
     && now+(next_event->elapsed/next_event->count)>after_occurance)
    make_new_thread();

  debug("%d threads, %d idle\n",num_threads,idle_threads);

  pthread_mutex_unlock(&thread_lock);

  debug("Returning event interval %u at %d.  Average is %u.  Next is at %d.\n",
	next_event->interval,(int)next_event->next_occurance,
	next_event->count?next_event->elapsed/next_event->count:0,
	(int)after_occurance);

  pthread_cleanup_pop(0);

  pthread_mutex_unlock(&event_lock);

  next_event->last_start=now;

  return next_event;
}

static void *
periodic_thread(void *foo)
{
  for(;;)
    {
      struct periodic_event_t *event;

      /* Get it */
      event=dequeue();

      pthread_cleanup_push(enqueue,event);

      /* Execute it */
      (*event->func)(event->arg);

      /* Give it back */
      pthread_cleanup_pop(1);
    }

  /* Never reached */
  return NULL;
}

struct periodic_event_t *
periodic_add(unsigned int interval,unsigned int flags,
	     void (*func)(void *),void *arg)
{
  struct periodic_event_t *event;

  event=calloc(1,sizeof(*event));
  if(!event)
    {
      errno=ENOMEM;
      return NULL;
    }

  event->interval=interval;
  event->func=func;
  event->arg=arg;

  if(flags&PERIODIC_DELAY)
    event->next_occurance=gettime()+interval;

  if(flags&PERIODIC_ONESHOT)
    event->flags.oneshot=1;

  pthread_mutex_lock(&event_lock);
  event->next=events;
  events=event;
  pthread_cond_broadcast(&event_cond);
  pthread_mutex_unlock(&event_lock);

  return event;
}

int
periodic_remove(struct periodic_event_t *remove)
{
  struct periodic_event_t *event,*last=NULL;

  pthread_mutex_lock(&event_lock);

  for(event=events;event;last=event,event=event->next)
    if(remove==event)
      {
	if(last)
	  last->next=event->next;
	else
	  events=event->next;

	free(event);

	pthread_cond_broadcast(&event_cond);
	break;
      }

  if(!event)
    {
      /* We didn't find it, which indicates that it might be running
	 right now.  Set the oneshot flag to make this into a oneshot
	 event, which will be deleted when the thread that is running
	 it tries to re-enqueue it. */

      remove->flags.oneshot=1;
    }

  pthread_mutex_unlock(&event_lock);

  return 0;
}

static void
prepare(void)
{
  pthread_mutex_lock(&event_lock);
}

static void
unprepare(void)
{
  pthread_mutex_unlock(&event_lock);
}

int
periodic_start(unsigned int flags)
{
  int err;
  pthread_condattr_t attr;

  if((err=pthread_condattr_init(&attr)))
    {
      errno=err;
      return -1;
    }

  /* Only use the monotonic clock if we can use it in both gettime and
     the pthreads condition variable. */
#if defined(HAVE_CLOCK_GETTIME) && defined(HAVE_PTHREAD_CONDATTR_SETCLOCK)

  if((err=pthread_condattr_setclock(&attr,CLOCK_MONOTONIC)))
    {
      pthread_condattr_destroy(&attr);
      errno=err;
      return -1;
    }
#endif /* HAVE_CLOCK_GETTIME && HAVE_PTHREAD_CONDATTR_SETCLOCK */

  if((err=pthread_cond_init(&event_cond,&attr)))
    {
      pthread_condattr_destroy(&attr);
      errno=err;
      return -1;
    }

  pthread_condattr_destroy(&attr);

  pthread_mutex_lock(&thread_lock);

  if(num_threads)
    {
      /* We're already running. */
      pthread_mutex_unlock(&thread_lock);
      errno=EBUSY;
      return -1;
    }

  global_flags=flags;

  err=pthread_atfork(prepare,unprepare,unprepare);
  if(err)
    {
      pthread_mutex_unlock(&thread_lock);
      errno=err;
      return -1;
    }

  if(flags&PERIODIC_NORETURN)
    {
      threads=malloc(sizeof(pthread_t));
      if(!threads)
	{
	  pthread_mutex_unlock(&thread_lock);
	  errno=ENOMEM;
	  return -1;
	}

      threads[0]=pthread_self();
      num_threads=1;
    }
  else
    {
      err=make_new_thread();
      if(err)
	{
	  pthread_mutex_unlock(&thread_lock);
	  errno=err;
	  return -1;
	}
    }

  pthread_mutex_unlock(&thread_lock);

  if(flags&PERIODIC_NORETURN)
    periodic_thread(NULL);

  return 0;
}

int
periodic_stop(unsigned int flags)
{
  unsigned int i;

  pthread_mutex_lock(&thread_lock);

  if(!num_threads)
    {
      /* We're not running. */
      pthread_mutex_unlock(&thread_lock);
      errno=EBUSY;
      return -1;
    }

  for(i=0;i<num_threads;i++)
    {
      if(!(flags&PERIODIC_WAIT))
	pthread_detach(threads[i]);

      pthread_cancel(threads[i]);
    }

  if(flags&PERIODIC_WAIT)
    for(i=0;i<num_threads;i++)
      pthread_join(threads[i],NULL);

  free(threads);
  num_threads=0;

  pthread_mutex_unlock(&thread_lock);

  return 0;
}

static void *
timewarp_thread(void *foo)
{
  time_t last_time=gettime();

  for(;;)
    {
      time_t now;
      unsigned int remaining=timewarp_interval;

      while(remaining)
	remaining=sleep(remaining);

      now=gettime();

      /* last_time+timewarp->interval is where we should be, if there
	 was no timewarp. */

      if(now>last_time+timewarp_interval+timewarp_warptime
	 || now<last_time+timewarp_interval-timewarp_warptime)
	{
	  struct periodic_event_t *event;

	  /* We've jumped more than warptime seconds. */

	  if(timewarp_func)
	    (timewarp_func)(timewarp_arg);

	  /* Wake everyone up and make them recalibrate. */

	  pthread_mutex_lock(&event_lock);

	  /* Recalculate everyone */

	  for(event=events;event;event=event->next)
	    {
	      event->next_occurance=now+event->interval;
	      event->last_start=0;
	      event->elapsed=0;
	      event->count=0;
	    }

	  pthread_cond_broadcast(&event_cond);
	  pthread_mutex_unlock(&event_lock);

	  /* This is because the timewarp_func may take a while to
	     execute */

	  now=gettime();
	}

      last_time=now;
    }

  /* Never reached */
  return NULL;
}

int
periodic_timewarp(unsigned int interval,unsigned int warptime,
		  void (*func)(void *),void *arg)
{
  if(interval)
    {
      int err;

      timewarp_interval=interval;
      timewarp_warptime=warptime;
      timewarp_func=func;
      timewarp_arg=arg;

      err=pthread_create(&timewarp,NULL,timewarp_thread,NULL);
      if(err==0)
	pthread_detach(timewarp);
      else
	{
	  errno=err;
	  return -1;
	}
    }

  return 0;
}
